123-job-jwt/proposal.md

```diff
@@ -0,0 +1,175 @@
+
+```
+
+This would represent a reasonable configuration. If it is determined that
+`auth_params` field in the `var_source.config.vault_config` schema should be
+re-used, then `auth_params: {role: "example", jwt: "((.CONCOURSE_CI_JWT))"}`
+might be a reasonable value, assuming that local environment variables can be
+referenced in such a way and we have pre-populated `CONCOURSE_CI_JWT` with the
+signed JWT during container pre-flight. However, this format would require more
+awareness on the part of the person configuring the pipeline, and the `jwt`
+field in the `auth_params` argument would always be the same, so it seems to
+make sense to abstract it in this case.
+
+Then, the call to the Vault API for the authentication step might look like the
+following:
+
+```sh
+#!/usr/bin/env bash
+auth_path=$(echo ${auth_path} | sed -E "s|^/*||g" | sed -E "s|/*$||g" |
+  sed -E "s|^auth/?||g" | sed -E "s|/?login$||g" | sed -E "s|/+|/|g"
+)
+resp=$(curl -s -X POST -H "X-Vault-Namespace: ${namespace}" \
+  --data @payload.json ${url}/v1/auth/${auth_path}/login
+)
+export VAULT_TOKEN=$(echo "${resp}" | jq -r ".auth.client_token")
+```
+
+or (forgive my golang pseudocode, rusty and untested):
+
+```golang
+// not dealing with errors or worrying about syntax, this is pseudocode
+// but please still make fun of my pseudocode, maybe I'll learn something
+func (ap *VarSources.Config.auth_path) FmtAuthPath() nil {
+  ap := strings.Trim(ap, "/")
+  ap := strings.TrimPrefix(ap, "auth/")
+  ap := strings.TrimSuffix(ap, "/login")
+  ap := path.Clean(ap)
+}
+type LoginData struct {
+  Role string `json:"role"`
+  JWT  string `json:"jwt"`
+}
+type LoginResp struct {
+  auth struct {
+    ClientToken string   `json:"client_token"`
+    Accessor    string   `json:"accessor"`
+    Policies    []string `json:"policies"`
+    Lease       int      `json:"lease_duration"`
+    Renewable   bool     `json:"renewable"`
+  } `json:"auth"`
+}
+func (c *VarSources.Config) VaultAuth() (token string, accessor string, err error) {
+  // I'm sure the vault package implements the auth method much better than this
+  client := &http.Client{}
+  uri := fmt.Sprintf("%s/v1/auth/%s/login", c.url, c.auth_path.FmtAuthPath())
+  jsonData := LoginData{
+    Role: c.auth_role,
+    JWT:  os.Getenv("CONCOURSE_CI_JWT")
+  }
+  var buf bytes.Buffer
+  err := json.NewEncoder(&buf).Encode(jsonData)
+  req, err := http.NewRequest("POST", uri, &buf)
+  req.Header.Add("X-Vault-Namespace", c.namespace)
+  resp, err := client.Do(req)
+  defer resp.Body.Close()
+  var dec LoginResp
+  dec = json.NewDecoder(&buf).Decode(resp)
+  return dec.auth.ClientToken, dec.auth.Accessor, nil
+}
+```
+
+## Open Questions
+
+- Where should the claims be aggregated and signing portion be completed? (atc?)
+- How should the signed JWT be delivered to the container runtime?
+- Where should the validation endpoint live? (web?)
+- Where should the signed JWT live? (Environment Variable + File?)
```

Comment:
Yes, it's Base 64 delimited by periods `.` so very safe to put into env vars. I think taking file storage off the table will simplify the RFC, and it could always be added later if someone has a use case that can't be filled by the env var for whatever reason.

Reply:
If nobody else has a comment here I will move this question to the answered questions section.